File: CustomCalloutView/ViewController.m
//
//  ViewController.m
//  CustomCalloutView
//
//  Created by 农盟 on 15/7/31.
//  Copyright (c) 2015年 农盟. All rights reserved.
//

#import "ViewController.h"
#import <MapKit/MapKit.h>
#import <CoreLocation/CoreLocation.h>
#import "YNBaseAnnotation.h"
#import "YNCallOutAnnotation.h"
#import "YNCallOutAnnotationView.h"
#import "YNLocationAnnotation.h"
#import "YNCallOutContentView.h"

@interface ViewController ()<MKMapViewDelegate>

@property (nonatomic, strong)  MKMapView *mapview;

@property (nonatomic, strong)  YNCallOutAnnotation *callOutAnnotation;
@property (nonatomic, strong)  YNCallOutAnnotationView *callOutAnnotationView;

@end

@implementation ViewController

#pragma mark - life cycle
- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self.view addSubview:self.mapview];
}

#pragma mark - MKMapViewDelegate

- (MKAnnotationView *)mapView:(MKMapView *)mapView viewForAnnotation:(id <MKAnnotation>)annotation {
    
    if ([annotation isKindOfClass:[YNBaseAnnotation class]]) {
        
        static NSString *identify = @"PINANNOTATIONVIEW";
        
        MKPinAnnotationView *pinAnnotationView = (MKPinAnnotationView *)[mapView dequeueReusableAnnotationViewWithIdentifier:identify];
        if (pinAnnotationView == nil) {
            pinAnnotationView = [[MKPinAnnotationView alloc] initWithAnnotation:annotation reuseIdentifier:identify];
        }
        pinAnnotationView.canShowCallout = NO;
        return pinAnnotationView;
        
    } else  if ([annotation isKindOfClass:[YNCallOutAnnotation class]]){
        
        static NSString *identify = @"CALLOUTANNOTATIONVIEW";
        
        YNCallOutAnnotationView *callOutAnnotationView = (YNCallOutAnnotationView *)[mapView dequeueReusableAnnotationViewWithIdentifier:identify];
        if (callOutAnnotationView == nil) {
            callOutAnnotationView = [[YNCallOutAnnotationView alloc] initWithAnnotation:annotation reuseIdentifier:identify];
        }
        
        callOutAnnotationView.alpha = 1.0;
        self.callOutAnnotationView = callOutAnnotationView;
        YNCallOutContentView *view = [[YNCallOutContentView alloc] initWithFrame:callOutAnnotationView.bounds];
        [callOutAnnotationView addSubview:view];
        
        return callOutAnnotationView;
        
    } else if ([annotation isKindOfClass:[YNLocationAnnotation class]]) {
        
        static NSString *identify = @"LOCATIONANNOTATIONVIEW";
        
        MKPinAnnotationView *pinAnnotationView = (MKPinAnnotationView *)[mapView dequeueReusableAnnotationViewWithIdentifier:identify];
        if (pinAnnotationView == nil) {
            pinAnnotationView = [[MKPinAnnotationView alloc] initWithAnnotation:annotation reuseIdentifier:identify];
        }
        pinAnnotationView.canShowCallout = YES;
        pinAnnotationView.image = [UIImage imageNamed:@"tarBar_nearby_off"];
        pinAnnotationView.calloutOffset = CGPointMake(-1, -2);
        return pinAnnotationView;
    }
    
    return nil;
    
}

- (void)mapView:(MKMapView *)mapView didSelectAnnotationView:(MKAnnotationView *)view {
    
    if ([view.annotation isKindOfClass:[YNBaseAnnotation class]]) {
        
        YNCallOutAnnotation *callOutAnnotation = [[YNCallOutAnnotation alloc] init];
        
        [callOutAnnotation setCoordinate:view.annotation.coordinate];
        [mapView addAnnotation:callOutAnnotation];
        self.callOutAnnotation = callOutAnnotation;
    }
    
}

- (void)mapView:(MKMapView *)mapView didDeselectAnnotationView:(MKAnnotationView *)view {
    
    if (self.callOutAnnotation) {
        
        [UIView animateWithDuration:0.4 animations:^{
            
            self.callOutAnnotationView.alpha = 0.0;
            
        } completion:^(BOOL finished) {
            
            [self.mapview removeAnnotation: self.callOutAnnotation];
            
            self.callOutAnnotation = nil;
        }];
        
    }
}

- (void)mapViewDidFinishRenderingMap:(MKMapView *)mapView fullyRendered:(BOOL)fullyRendered {
    
    CLLocationCoordinate2D centerCoordinate1 = CLLocationCoordinate2DMake(34.42, 113.42);
    
    YNBaseAnnotation *baseAnnotation = [[YNBaseAnnotation alloc] init];
    [baseAnnotation setCoordinate:centerCoordinate1];
    [_mapview addAnnotation:baseAnnotation];
    
    CLLocationCoordinate2D locationCenterCoordinate = CLLocationCoordinate2DMake(34.10, 113.30);
    YNLocationAnnotation *locationAnnotation = [[YNLocationAnnotation alloc] init];
    [locationAnnotation setCoordinate:locationCenterCoordinate];
    locationAnnotation.title = @"hello coder";
    locationAnnotation.subtitle = @"时光穿梭到未来的2080年看看";
    [self.mapview addAnnotation:locationAnnotation];
    [self.mapview selectAnnotation:locationAnnotation animated:YES];
}

#pragma getters and setters
- (MKMapView *)mapview {
    if (_mapview == nil) {
        _mapview = [[MKMapView alloc] initWithFrame:self.view.bounds];
        _mapview.mapType = MKMapTypeStandard;
        _mapview.userTrackingMode = MKUserTrackingModeFollow;
        _mapview.delegate = self;
        CLLocationCoordinate2D centerCoordinate1 = CLLocationCoordinate2DMake(34.42, 113.42);
        
        MKCoordinateSpan span = MKCoordinateSpanMake(1, 1);
        MKCoordinateRegion regin = MKCoordinateRegionMake(centerCoordinate1, span);
        [_mapview setRegion:regin animated:YES];
        
    }
    return _mapview;
}


@end

File: SwiftVwesion/SwiftVwesion/ViewController.swift
//
//  ViewController.swift
//  SwiftVwesion
//
//  Created by 农盟 on 15/8/19.
//  Copyright (c) 2015年 农盟. All rights reserved.
//

import UIKit
import MapKit

class ViewController: UIViewController, MKMapViewDelegate {

    var callOutAnnotationView: YNCallOutAnnotationView?
    var callOutAnnotation: YNCallOutAnnotation?
    
    var isDeleteAnnotation: Bool = false
    
    lazy var mapView: MKMapView = {
        var tempMapView = MKMapView(frame: self.view.bounds)
        tempMapView.showsUserLocation = true
        tempMapView.userTrackingMode = MKUserTrackingMode.Follow
        tempMapView.mapType = MKMapType.Standard
        tempMapView.delegate = self
        return tempMapView
        }()
    
    var coordinate: CLLocationCoordinate2D {
        
        return CLLocationCoordinate2DMake(43.7736, 113.718)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.view.addSubview(self.mapView)
        let span: MKCoordinateSpan = MKCoordinateSpanMake(0.003739, 0.003201)
        let regin: MKCoordinateRegion = MKCoordinateRegionMake(self.coordinate, span)
        self.mapView.setRegion(regin, animated: true)
        
        addEnterpriseAnnotation()
    }
    
    
    //MARK: - MKMapViewDelegate
    func mapView(mapView: MKMapView, viewForAnnotation annotation: MKAnnotation) -> MKAnnotationView? {
        
        if annotation is YNBaseAnnotation {
            
            let identify = "PINANNOTATIONVIEW"
            var pinAnnotationView: MKPinAnnotationView? = mapView.dequeueReusableAnnotationViewWithIdentifier(identify) as? MKPinAnnotationView
            if pinAnnotationView == nil {
                pinAnnotationView = MKPinAnnotationView(annotation: annotation, reuseIdentifier: identify)
                
            }
            
            pinAnnotationView?.canShowCallout = false
            pinAnnotationView?.pinColor = MKPinAnnotationColor.Purple
            return pinAnnotationView
            
        } else if annotation is YNCallOutAnnotation {
            
            let callOutidentify = "CALLOUTANNOTATIONVIEW"
            var callOutAnnotationView: YNCallOutAnnotationView? = mapView.dequeueReusableAnnotationViewWithIdentifier(callOutidentify) as? YNCallOutAnnotationView
            if callOutAnnotationView == nil {
                
                callOutAnnotationView = YNCallOutAnnotationView(annotation: annotation, reuseIdentifier: callOutidentify)
            }
            
            callOutAnnotationView!.alpha = 1.0
            self.callOutAnnotationView = callOutAnnotationView
            
            let contentView: ContentView = ContentView(frame: callOutAnnotationView!.bounds)
            
            callOutAnnotationView?.addSubview(contentView)
            
            return callOutAnnotationView
        }
        
        
        return nil
    }
    
    func mapView(mapView: MKMapView, didSelectAnnotationView view: MKAnnotationView) {
        
        if view.annotation is YNBaseAnnotation {
            
            self.isDeleteAnnotation = false
//            let baseAnnotation: YNBaseAnnotation = view.annotation as! YNBaseAnnotation
            
            //TODO: 纪录 baseAnnotation.index
            
            let callOutAnnotation = YNCallOutAnnotation(coordinate: view.annotation!.coordinate)
            mapView.addAnnotation(callOutAnnotation)
            self.callOutAnnotation = callOutAnnotation
        }
        
    }
    
    func mapView(mapView: MKMapView, didDeselectAnnotationView view: MKAnnotationView) {
        
        if let _ = self.callOutAnnotation {
            
            self.isDeleteAnnotation = true
            
            UIView.animateWithDuration(0.3 , animations: { () -> Void in
                
                self.callOutAnnotationView!.alpha = 0.0
                
                }, completion: { (finished) -> Void in
                    
                    if self.isDeleteAnnotation {
                        
                        if let _ = self.callOutAnnotation {
                        
                            self.mapView.removeAnnotation(self.callOutAnnotation!)
                            self.callOutAnnotation = nil
                            
                        }
                        
                        
                    }
                    
            })
            
            
        }
        
    }
    
    
    func addEnterpriseAnnotation() {
        
        let coordinate1: CLLocationCoordinate2D = CLLocationCoordinate2DMake(self.coordinate.latitude - 0.001, self.coordinate.longitude )
        let baseAnnotation1: YNBaseAnnotation = YNBaseAnnotation(coordinate: coordinate1)
        baseAnnotation1.index = 1
        
        let coordinate2: CLLocationCoordinate2D = CLLocationCoordinate2DMake(self.coordinate.latitude + 0.001, self.coordinate.longitude )
        let baseAnnotation2: YNBaseAnnotation = YNBaseAnnotation(coordinate: coordinate2)
        baseAnnotation2.index = 2
        
        self.mapView.addAnnotations([baseAnnotation1, baseAnnotation2])
        
    }


}

